torchqdynamics/odeint.py

```diff
@@ -16,8 +16,37 @@
 
 
 class ForwardQSolver(ABC):
-    def __init__(self, solver_options: SolverOption):
-        self.options = solver_options
+    def __init__(
+        self, options: SolverOption, y0: Tensor, exp_ops: Tensor, t_save: Tensor
+    ):
+        self.options = options
+        self.exp_ops = exp_ops
+        self.t_save = t_save
+
+        self.save_counter = 0
+
+        # initialize save tensors
+        batch_sizes, (m, n) = y0.shape[:-2], y0.shape[-2:]
+
+        if self.options.save_states:
+            # y_save: (..., len(t_save), m, n)
+            self.y_save = torch.zeros(
+                *batch_sizes, len(self.t_save), m, n, dtype=y0.dtype, device=y0.device
+            )
+
+        if len(self.exp_ops) > 0:
+            # exp_save: (..., len(exp_ops), len(t_save))
+            self.exp_save = torch.zeros(
+                *batch_sizes,
+                len(self.exp_ops),
+                len(self.t_save),
+                dtype=y0.dtype,
+                device=y0.device,
+            )
+        else:
+            self.exp_save = torch.empty(
+                *batch_sizes, len(self.exp_ops), dtype=y0.dtype, device=y0.device
+            )
```

Comment:
Unless there's a specific reason for this, I think `exp_save` should be None if no `exp_ops` are passed.